types/query/pagination.go

```diff
@@ -0,0 +1,93 @@
+package query
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// defaultLimit is the default `limit` for queries
+// if the `limit` is not supplied, paginate will use `defaultLimit`
+const defaultLimit = 100
+
+// Paginate does pagination of all the results in the PrefixStore based on the
+// provided PageRequest. onResult should be used to do actual unmarshaling.
+func Paginate(
+	prefixStore types.KVStore,
+	req *PageRequest,
+	onResult func(key []byte, value []byte) error,
+) (*PageResponse, error) {
+	offset := req.Offset
+	key := req.Key
+	limit := req.Limit
+	countTotal := req.CountTotal
+
+	if offset > 0 && key != nil {
+		return nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
+	}
+
+	if limit == 0 {
+		limit = defaultLimit
+
+		// count total results when the limit is zero/not supplied
+		countTotal = true
+	}
+
+	if len(key) != 0 {
+		iterator := prefixStore.Iterator(key, nil)
+		defer iterator.Close()
+
+		var count uint64
+		var nextKey []byte
+
+		for ; iterator.Valid(); iterator.Next() {
+			if count == limit {
+				nextKey = iterator.Key()
+				break
+			}
+
+			err := onResult(iterator.Key(), iterator.Value())
+			if err != nil {
+				return nil, err
+			}
+
+			count++
+		}
+
+		return &PageResponse{
+			NextKey: nextKey,
+		}, nil
+	}
+
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	end := offset + limit
+
+	var count uint64
+	var nextKey []byte
+
+	for ; iterator.Valid(); iterator.Next() {
+		count++
+
+		if count <= offset {
+			continue
+		}
+		if count <= end {
+			err := onResult(iterator.Key(), iterator.Value())
+			if err != nil {
+				return nil, err
+			}
+		} else if !countTotal {
+			nextKey = iterator.Key()
+			break
+		}
```

Comment:
I didn't catch this in my review but this isn't quite right. We should set `nextKey` here regardless of whether or not `countTotal` is set.

The correct code would be something like:
```suggestion
		} else count == end + 1 {
 			nextKey = iterator.Key()
 			
 			if !countTotal {
 			  break
 			}
 		}
```

Reply:
Addressed it here: https://github.com/cosmos/cosmos-sdk/pull/6514/commits/fe4aff1a5420960586ae34f947ca8f4372b26ac7